zerver/tests/test_markdown_thumbnail.py

```diff
@@ -0,0 +1,309 @@
+        ensure_thumbnails(ImageAttachment.objects.get(path_id=second_path_id))
+        self.assert_message_content_is(
+            message_id,
+            (
+                f'<p><a href="/user_uploads/{first_path_id}">first image</a><br>\n'
+                f'<a href="/user_uploads/{second_path_id}">second image</a></p>\n'
+                f'<div class="message_inline_image"><a href="/user_uploads/{first_path_id}" title="first image">'
+                '<img class="image-loading-placeholder" src="/static/images/loading/loader-black.svg"></a></div>'
+                f'<div class="message_inline_image"><a href="/user_uploads/{second_path_id}" title="second image">'
+                f'<img src="/user_uploads/thumbnail/{second_path_id}/840x560.webp"></a></div>'
+            ),
+        )
+
+        # Finish the other thumbnail
+        ensure_thumbnails(ImageAttachment.objects.get(path_id=first_path_id))
+        self.assert_message_content_is(
+            message_id,
+            (
+                f'<p><a href="/user_uploads/{first_path_id}">first image</a><br>\n'
+                f'<a href="/user_uploads/{second_path_id}">second image</a></p>\n'
+                f'<div class="message_inline_image"><a href="/user_uploads/{first_path_id}" title="first image">'
+                f'<img src="/user_uploads/thumbnail/{first_path_id}/840x560.webp"></a></div>'
+                f'<div class="message_inline_image"><a href="/user_uploads/{second_path_id}" title="second image">'
+                f'<img src="/user_uploads/thumbnail/{second_path_id}/840x560.webp"></a></div>'
+            ),
+        )
+
+    def test_thumbnail_of_deleted(self) -> None:
+        sender_user_profile = self.example_user("othello")
+        path_id = self.upload_image("img.png")
+        message_id = self.send_message_content(f"[image](/user_uploads/{path_id})")
+
+        # Delete the message
+        do_delete_messages(
+            sender_user_profile.realm, [Message.objects.get(id=message_id)], acting_user=None
+        )
+
+        # There is still an ImageAttachment row
+        self.assertFalse(Attachment.objects.filter(path_id=path_id).exists())
+        self.assertTrue(ArchivedAttachment.objects.filter(path_id=path_id).exists())
+        self.assertTrue(ImageAttachment.objects.filter(path_id=path_id).exists())
+
+        # Completing rendering after it is deleted should work, and
+        # update the rendered content in the archived message
+        ensure_thumbnails(ImageAttachment.objects.get(path_id=path_id))
+        expected = (
+            f'<p><a href="/user_uploads/{path_id}">image</a></p>\n'
+            f'<div class="message_inline_image"><a href="/user_uploads/{path_id}" title="image">'
+            f'<img src="/user_uploads/thumbnail/{path_id}/840x560.webp"></a></div>'
+        )
+        self.assertEqual(
+            ArchivedMessage.objects.get(id=message_id).rendered_content,
+            expected,
+        )
```

Comment:
Should we expect this test to confirm that undeleting the message "works", or that vacuuming deletes the `ImageAttachment`?

Reply:
The vacuuming tests are already in `zerver/tests/test_delete_unclaimed_attachments.py` added in 2e38f426f4ed2f2fa5e82b13ff2ec6afd22c4b11.

Since deleting doesn't touch ImageAttachments in any way, there's nothing for it to do when undeleting.

Reply:
OK, cool, maybe just add a comment here noting the vacuuming is tested in `test_delete_unclaimed_attachments` would be nice.